components/renku_data_services/connected_services/provider_adapters.py

```diff
@@ -232,12 +235,86 @@ def api_validate_account_response(self, response: Response) -> models.ConnectedA
-    ProviderKind.gitlab: GitLabAdapter,
-    ProviderKind.github: GitHubAdapter,
-    ProviderKind.drive: GoogleDriveAdapter,
-    ProviderKind.onedrive: OneDriveAdapter,
-    ProviderKind.dropbox: DropboxAdapter,
+class GenericOidcAdapter(ProviderAdapter):
+    """Adapter for generic OpenID Connect clients."""
+
+    _httpx_client: Client | None = None
+    _configurations: dict[str, dict[str, str]] = dict()
+
+    def __init__(self, client_url: str, oidc_issuer_url: str, **kwargs: Any) -> None:
+        super().__init__(client_url, **kwargs)
+        self.oidc_issuer_url = oidc_issuer_url
+
+    @property
+    def authorization_url(self) -> str:
+        """The authorization URL for the OAuth2 protocol."""
+        return self.__get_configuration()["authorization_endpoint"]
+
+    @property
+    def token_endpoint_url(self) -> str:
+        """The token endpoint URL for the OAuth2 protocol."""
+        return self.__get_configuration()["token_endpoint"]
+
+    @property
+    def api_url(self) -> str:
+        """The URL used for API calls on the Resource Server."""
+        return self.client_url
+
+    @property
+    def user_info_endpoint(self) -> str:
+        """The URL of the user info endpoint."""
+        return self.__get_configuration()["userinfo_endpoint"]
+
+    @user_info_endpoint.setter
+    def user_info_endpoint(self, value: str) -> None:
+        """Setter for user_info_endpoint."""
+        raise errors.ProgrammingError(message="Cannot set user_info_endpoint.")
+
+    def api_validate_account_response(self, response: Response) -> models.ConnectedAccount:
+        """Validates and returns the connected account response from the Resource Server."""
+        return external_models.GenericOIDCConnectedAccount.model_validate(response.json()).to_connected_account()
+
+    def __get_configuration(self) -> dict[str, str]:
+        config = self.__get_configurations().get(self.oidc_issuer_url, None)
+        if config is None:
+            config = self.__discover()
+            self.__get_configurations()[self.oidc_issuer_url] = config
+        return config
+
+    def __discover(self) -> dict[str, str]:
+        issuer_url = self.oidc_issuer_url
+        if not issuer_url:
+            raise errors.ValidationError(message="Issuer URL not configured for generic OIDC client.")
+        issuer_url = issuer_url if issuer_url.endswith("/") else issuer_url + "/"
+        request_url = urljoin(issuer_url, ".well-known/openid-configuration")
+        res = self.__get_httpx_client().get(request_url, headers=[("Accept", "application/json")])
+        if res.status_code != 200:
+            raise errors.ProgrammingError(message=f"Could not read OIDC issuer configuration from {request_url}")
+        res_json = res.json()
+        return {
+            "authorization_endpoint": res_json["authorization_endpoint"],
+            "token_endpoint": res_json["token_endpoint"],
+            "userinfo_endpoint": res_json["userinfo_endpoint"],
+        }
```

Comment:
question: have you checked if we already use a library that can just do this instead of us having to code it ourselves?

Reply:
I would rather not, considering that the OAuth 2.0 library we use has lead to spaghetti code and race conditions.

This is relatively simple to interpret and check against the relevant documentation. I can add a link to it here so that it is easy to maintain in the future.